### ISSUE TITLE
provide option to include private and special members

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#176]
it's pretty standard practice in API documentation to include special methods such as:

```
__add__, __mul__, __div__
```

because they are essential to conveying the purpose and usage of the class. additional options should be added along the lines of "undoc-members", for including special methods:

```
__foo__
```

and private methods:

```
_foo, __foo, _foo_
```

it's important to have individual control over these since it is often the case that private members should be skipped while special members should be documented.

```
.. autoclass:: Foo
  :members:
  :private-members:
  :special-members:
```

NOTE:
i am aware that "autodoc-skip-member" event can be used for this purpose, but 99&#37; of us are not willing to write custom modules and callbacks to accomplish something as simple and essential as including private members -- it should just work.  to my knowledge, the sphinx docs do not even provide examples of how one would go about hooking up an event callback.  it is unrealistic to expect the average user to delve into this magic to accomplish something so fundamental. 

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/176
- Originally reported by: Chad Dombrova
- Originally created at: 2009-05-19T23:33:13.275

**Comment by shimizukawa** [comment]
_From Anonymous on 2010-05-28 18:59:35+00:00_

+1. I am surprised there is not more demand for this--I've been trying to find autodoc-skip-member examples myself with no success. Sphinx can document the **init** method with the autoclass_content setting, but that's not ideal and frankly seems like a hack; there are many other special (non-private) methods that the API user will want to know about, like **getitem**, **getattr**, **gt**, **lt** and so on in addition to the ones chadrik mentioned above. It would be great if I could just put something like "include_private=True" in my conf.py to include these in the documentation. Right now my only alternative seems to be to include explicit autofunction directives for all the "private" methods I want to include.

**Comment by shimizukawa** [comment]
_From Éric Araujo on 2010-10-28 23:28:23+00:00_

Not sure whether this is a dupe of #520.  Copying my comment: IMO magic methods don’t have to be documented for a specific class since they’re documented once in the Python docs.

**Comment by shimizukawa** [comment]
_From Chad Dombrova on 2010-10-29 07:21:33+00:00_

yes, but magic members can be used to do different things on different classes.  certainly the python docs for these methods are not universally applicable.  for example, on a vector class **pow** might do cross product.  and for magic methods that take an input, like **mul**, **div**, etc, it's very necessary to document what **type** of object it works with on a particular class.  the docs might also provide examples of how to use the special methods in the context of the class being documented.  especially for classes that make heavy use of the numeric magic methods, not documenting all of this is a huge omission.

also, this feature request covers private methods as well, which are sometimes necessary to document.  i'm not suggesting these two features should be on by default, but they should exist.

**Comment by shimizukawa** [comment]
_From Kevin Horn on 2010-11-02 20:19:15+00:00_

+1, would love to have this functionality

**Comment by shimizukawa** [comment]
_From Max Cantor on 2010-11-30 21:56:48+00:00_

= +1! =

== This is vital for internal documentation. ==
I'm trying to use Sphinx to generate //internal// documentation for a huge codebase.  

=== I'm not just documenting the public API. ===
It is not for public consumption; it's meant to be read and used by our internal developers, particularly with new hires or when people switch teams and projects.  

== //autodoc-skip-member// doesn't work ==
Right now, you can't even use //autodoc-skip-member//, because the event **isn't fired** by a lone //:members:// directive: you //must// explicitly list members, which kind of defeats the purpose.

**Comment by shimizukawa** [comment]
_From Max Cantor on 2010-11-30 21:57:47+00:00_

I'm not sure why Mr. Horn removed this from the "autodoc" component, but I'm adding it back since it seems appropriate.

**Comment by shimizukawa** [comment]
_From Kevin Horn on 2010-12-01 20:19:28+00:00_

[Max](https://bitbucket.org/Max):

I certainly had no intention of changing the component for this issue.  Either I did it accidentally or bitbucket's form did it for me.

In any case, thanks for fixing it.

**Comment by shimizukawa** [comment]
_From Max Cantor on 2010-12-02 02:53:57+00:00_

Hello all.  I have attached a patch which enables this functionality.  This didn't seem like a big enough change to warrant forking the entire repository, but I couldn't figure out the right way to submit a plain old patch file.  So... here it is!  I also "updated" the tests, though only cursorily; the existing tests are a little hairy, and did not seem to cover the member skipping logic.

I am presumptuously assigning this issue to Mr. Brandl; not sure who is actually supposed to vet patches.  Thanks!

[kevin](https://bitbucket.org/Kevin): No prob!

Also, it seems I spoke too soon when I blaringly declared "**//autodoc-skip-member// doesn't work**" before; it works fine!  I erroneously thought that the //sphinx-build// script counted "conf.py" as a dependency.  I interpreted the lack of difference in my resulting docs as an indication that //autodoc-skip-member// wasn't working as expected, when in fact, it simply wasn't being emitted because I hadn't deleted my //_build// directory.
